### PR TITLE
Add nullptr checks to efi DeviceHandle and FilePath

### LINKED ISSUE #29331: Systemd-stub fails to boot if booted by grub on non-efi readable partition
### systemd version the issue has been seen with

254.4-2

### Used distribution

Arch

### Linux kernel version used

6.5.4-arch2-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-stub

### Expected behaviour you didn't see

I have a UKI generated by ukify, sitting on an ext4 formatted boot partition. As this partition is not readable by my uefi, i would expect systemd-stub to not check for additional files and just boot Linux.

### Unexpected behaviour you saw

 When my UKI is called by grub, grub passes `null` as the `DeviceHandle` in the loaded image protocol. `systemd-stub` asserts the device handle to be non-null, causing the boot to fail, and shows the very non-helpful error message
```
systemd-boot: Assertion 'device' failed at ../systemd-stable/src/boot/efi/part-discovery.c:251@partition_open, halting.
```

### Steps to reproduce the problem

Use the grub `chainloader` command to boot a UKI containing `systemd-stub`, which is loaded from an ext4 partition.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by lufsc**
Fixes #29331.

### CODE REVIEW

File: src/boot/efi/stub.c [hunk 1]
```diff
@@ -394,9 +394,11 @@ static EFI_STATUS run(EFI_HANDLE image) {
             !FLAGS_SET(loader_features, EFI_LOADER_FEATURE_RANDOM_SEED)) {
                 _cleanup_(file_closep) EFI_FILE *esp_dir = NULL;
 
-                err = partition_open(MAKE_GUID_PTR(ESP), loaded_image->DeviceHandle, NULL, &esp_dir);
-                if (err == EFI_SUCCESS) /* Non-fatal on failure, so that we still boot without it. */
-                        (void) process_random_seed(esp_dir);
+                if (loaded_image->DeviceHandle) {
```
Comment (yuwata) [inline src/boot/efi/stub.c]: Please merge this with the above conditions `efivar_get_uint64_le(...)` and `!FLAGS_SET(...)`.

File: src/boot/efi/stub.c [hunk 2]
```diff
@@ -482,17 +482,19 @@ static EFI_STATUS run(EFI_HANDLE image) {
...
-                        image,
-                        loaded_image,
-                        dropin_dir,
-                        uname,
-                        &m,
-                        &cmdline);
-        if (err != EFI_SUCCESS)
-                log_error_status(err, "Error loading UKI-specific addons, ignoring: %m");
-        parameters_measured = parameters_measured < 0 ? m : (parameters_measured && m);
+        if (loaded_image->FilePath) {
```
Comment (poettering) [inline src/boot/efi/stub.c]: can you add a comment here: `/* handle case where not backed by proper file system */` or so?

File: src/boot/efi/stub.c [hunk 3]
```diff
@@ -390,8 +390,8 @@ static EFI_STATUS run(EFI_HANDLE image) {
         if (err != EFI_SUCCESS)
                 return log_error_status(err, "Error getting a LoadedImageProtocol handle: %m");
 
-        if (efivar_get_uint64_le(MAKE_GUID_PTR(LOADER), u"LoaderFeatures", &loader_features) != EFI_SUCCESS ||
-            !FLAGS_SET(loader_features, EFI_LOADER_FEATURE_RANDOM_SEED)) {
+        if (loaded_image->DeviceHandle && (efivar_get_uint64_le(MAKE_GUID_PTR(LOADER), u"LoaderFeatures", &loader_features) != EFI_SUCCESS ||
+            !FLAGS_SET(loader_features, EFI_LOADER_FEATURE_RANDOM_SEED))) {
```
Comment (poettering) [inline src/boot/efi/stub.c:525]: can you add a comment here, why we need to check this explicitly, i.e. maybe format like this:

```c
if (loaded_image->DeviceHandle && /* handle case where image is not backed by proper file system */
        (efivar_get_uint64_le(…
```

**Comment by poettering** [comment]
codewise looks good, but please add comments, that explain these checks, as suggested.

also, please explain in the commit msg what is going on here.

**Pull request merged**